# Predict Red Wine Quality with Supervised Machine Learning

In [1]:
#Importing required packages.
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.ensemble import RandomForestClassifier
from sklearn.svm import SVC
from sklearn.linear_model import SGDClassifier
from sklearn.metrics import confusion_matrix, classification_report
from sklearn.preprocessing import StandardScaler, LabelEncoder
from sklearn.model_selection import train_test_split, GridSearchCV, cross_val_score
%matplotlib inline

In [2]:
redwine = pd.read_csv(r'C:\Users\nikki\OneDrive\Pictures\DATASC~1\PERSON~1\REDWIN~1\WINEQU~1.CSV')
redwine.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [3]:
redwine.shape

(1599, 12)

# Data Wrangling

In [4]:
#Making binary classificaion for the response variable.
#Dividing wine as good and bad by giving the limit for the quality
bins = (2, 6.5, 8)
group_names = ['bad', 'good']
redwine['quality'] = pd.cut(redwine['quality'], bins = bins, labels = group_names)

In [5]:
redwine.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol quality  
0      9.4     bad  
1      9.8     bad  
2      9.8     bad  
3      9.8     bad  
4      9.4     bad

In [6]:
redwine.quality.value_counts()

bad     1382
good     217
Name: quality, dtype: int64

In [7]:
#Now lets assign a labels to our quality variable
label_quality = LabelEncoder()
#Bad becomes 0 and good becomes 1 
redwine['quality'] = label_quality.fit_transform(redwine['quality'])

In [8]:
redwine.quality.value_counts()

0    1382
1     217
Name: quality, dtype: int64

In [9]:
#Now seperate the dataset as response variable and feature variabes
x = redwine.drop('quality', axis=1)
y = redwine['quality']

# Train/Test Split

In [10]:
#Train and Test splitting of data 
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size = 0.2, random_state = 42)

In [11]:
#Applying Standard scaling to get optimized result
sc = StandardScaler()

In [12]:
x_train = sc.fit_transform(x_train)
x_test = sc.fit_transform(x_test)

In [13]:
# See the shape of your data
print(x_train.shape, y_train.shape)
print(x_test.shape, y_test.shape)

(1279, 11) (1279,)
(320, 11) (320,)


# Random Forest Classifier

In [14]:
rfc = RandomForestClassifier(n_estimators=200)
rfc.fit(x_train, y_train)
pred_rfc = rfc.predict(x_test)

In [15]:
#Let's see how our model performed
print(classification_report(y_test, pred_rfc))

              precision    recall  f1-score   support

           0       0.90      0.97      0.93       273
           1       0.67      0.34      0.45        47

    accuracy                           0.88       320
   macro avg       0.78      0.66      0.69       320
weighted avg       0.86      0.88      0.86       320



## Random Forest gives the accuracy of 88%

In [16]:
#Confusion matrix for the random forest classification
print(confusion_matrix(y_test, pred_rfc))

[[265   8]
 [ 31  16]]


# Stochastic Gradient Decent Classifier

In [17]:
sgd = SGDClassifier(penalty=None)
sgd.fit(x_train, y_train)
pred_sgd = sgd.predict(x_test)

In [18]:
print(classification_report(y_test, pred_sgd))

              precision    recall  f1-score   support

           0       0.87      0.91      0.89       273
           1       0.29      0.21      0.25        47

    accuracy                           0.81       320
   macro avg       0.58      0.56      0.57       320
weighted avg       0.79      0.81      0.80       320



## 84% accuracy using stochastic gradient descent classifier

# Cross Validation

In [19]:
#Now lets try to do some evaluation for random forest model using cross validation.
rfc_eval = cross_val_score(estimator = rfc, X = x_train, y = y_train, cv = 10)
rfc_eval.mean()

0.9124569389763779

## Random forest accuracy increases from 88% to 91 % using cross validation score